src/libs/actions/Report.js

```diff
@@ -168,6 +168,51 @@ function getChatReportName(fullReport, chatType) {
         .join(', ');
 }
 
+/**
+ * @param {Number} reportID
+ * @returns {Number}
+ */
+function getNumberOfOptimisticActions(reportID) {
+    const optimisticActionIds = lodashGet(allReports, [reportID, 'optimisticReportActionIDs'], []);
+    return _.filter(optimisticActionIds, actionId => ReportActions.doesOptimisticReportActionExist(reportID, actionId)).length;
+}
+
+/**
+ * If onyx local report data is newer than fetched report then return updated version of simplifiedReport
+ * Local report data could be newer than the fetched report because of a pusher new comment event coming while the fetching report request hasn't been completed yet
+ *
+ * @param {Object} simplifiedReport, simplifiedReport of Report Data
+ * @returns {Object}
+ */
+function checkAndUpdateSimplifiedReport(simplifiedReport) {
+    const currentReport = lodashGet(allReports, simplifiedReport.reportID, null);
+    if (!currentReport) {
+        return simplifiedReport;
+    }
+    const simplifiedReportMaxSeqNumber = lodashGet(simplifiedReport, 'maxSequenceNumber', 0);
+    const currentReportMaxSeqNumber = lodashGet(currentReport, 'maxSequenceNumber', 0);
+
+    // Exclude optimisticActionIds from maxSequenceNumber of current local report when comparing maxSequenceNumber of fetched report and local onyx report.
+    // Optimistic report actions is a new comment not yet sent to the server.
+    const currentMaxSeqNumberWithoutOptimisticActions = currentReportMaxSeqNumber - getNumberOfOptimisticActions(currentReport.reportID);
+    if (simplifiedReportMaxSeqNumber >= currentMaxSeqNumberWithoutOptimisticActions) {
+        return simplifiedReport;
+    }
+
+    const lastReadSequenceNumber = lastReadSequenceNumbers[simplifiedReport.reportID];
+    const unreadActionCount = currentMaxSeqNumberWithoutOptimisticActions - lastReadSequenceNumber - ReportActions.getDeletedCommentsCount(simplifiedReport.reportID, lastReadSequenceNumber);
```

Comment:
I will add a proper explanation. 

> All of this extra handling makes me wonder why we are not just sending the "unread count" from the server when fetching the report.


Hmm. it could be an optimization. I guess it is like this from the start. One reason could be that every user in the report will have a different unread count but this can also be handled.

Even if the report is sending the unread count we can't use that here. The whole point is that there is a delay in receiving the report data than pusher sending new messages. That's why we are doing this operation.